### ISSUE TITLE
For some blocks, Batch RPC return only some replies, or the replies have null/mismatched fields

### ISSUE THREAD

**Issue by wizzard0** [ethereum/go-ethereum#23132]
#### System information

Geth version: 1.10.1, 1.10.4
OS & Version: Win10, Ubuntu 18, Ubuntu 20

#### Expected behaviour
For the batch call over websocket, replies for all requests, for all blocks are consistent.

#### Actual behaviour
- The received JSON contains NULLs in place of some replies, or NULL/mismatched data fields.
- Fails to deserialize into strongly typed data structures if one uses Golang RPC client
Notes:
- Does not look related to chain reorgs, as no orphan headers are received before/after the bug is observed, and no reorgs are registered on Etherscan
- Does not reproduce with regular (non-batch) RPC calls.

#### Steps to reproduce the behaviour
1. Subscribe to newHeads
2. For each block header, submit `eth_getTransactionReceipt` or `debug_traceTransaction`  for every transaction.
3. Optional: Add logging of raw resp.Result in `rpc/client.go` right before json.Unmarshal (https://github.com/ethereum/go-ethereum/blob/master/rpc/client.go#L391)
4. Wait about 100-300 blocks
5. Observe the request with a lot of JSON unmarshal errors, observe NULLs/garbage in resp.result bytes instead of expected replies
#### Backtrace
N/A

When submitting logs: please submit them as text and not screenshots.

**Comment by ligi** [comment]
Which network are you on?

**Comment by wizzard0** [comment]
mainnet

**Comment by karalabe** [comment]
Our hunch is that it might be mini 1 block "reorgs". E.g. when a new block becomes canonical in place of an equal height old block. It might be a data race in that case: header gets announced, you request receipts, but by that time another header gets swapped in, so some of the transactions might be "undone".

Could you check if those txs that do get sent back, whether the inclusion header hash matches the one you requested? Could you extend your tester so that if you see a `null`, you double check if a new header became canonical?

---

If you have a snippet you could give us to repro exactly, that would help understand it better.

**Comment by wizzard0** [comment]
Yeah the header hash matches, reorgs were the first thing I suspected.

But I receive header 0xaaa, then notification for header 0xbbb, request receipts for 0xbbb, get like half of them (randomly), all referring to block 0xbbb, then receive notification for header 0xccc which references 0xbbb as parent, and nothing else.

If those were reorgs, then I assume I should have received more header notifications?

With tester, do you mean the client code, or try to patch to the node itself?

**Comment by pinkiebell** [comment]
I observe similar behaviour with a light-mode enabled node. Block hashes and transaction list etc. are all fine but if I invoke `eth_getTransactionByHash...` I sometimes receive the tx object with `input` being `null`.
But this is not easy to reproduce. Trying the call again may return the '0x...' data of the transaction - OR - the field stays 'null' unless the node is restarted.

Must have something todo with error handing/caching in the LES layer and maybe bad peers.
